File: go/test/ipn_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/world-time-sdk/go"
	"github.com/voxgig-sdk/world-time-sdk/go/core"

	vs "github.com/voxgig-sdk/world-time-sdk/go/utility/struct"
)

func TestIpnEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Ipn(nil)
		if ent == nil {
			t.Fatal("expected non-nil IpnEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := ipnBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "ipn." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set WORLDTIME_TEST_IPN_ENTID JSON to run live")
			return
		}
		// Bootstrap entity data from existing test data (no create step in flow).
		ipnRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.ipn", setup.data)))
		var ipnRef01Data map[string]any
		if len(ipnRef01DataRaw) > 0 {
			ipnRef01Data = core.ToMapAny(ipnRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = ipnRef01Data

	})
}

func ipnBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "ipn", "IpnTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read ipn test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse ipn test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"ipn01", "ipn02", "ipn03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("WORLDTIME_TEST_IPN_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"WORLDTIME_TEST_IPN_ENTID": idmap,
		"WORLDTIME_TEST_LIVE":      "FALSE",
		"WORLDTIME_TEST_EXPLAIN":   "FALSE",
	})

	idmapResolved := core.ToMapAny(env["WORLDTIME_TEST_IPN_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["WORLDTIME_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
			},
			extra,
		})
		client = sdk.NewWorldTimeSDK(core.ToMapAny(mergedOpts))
	}

	live := env["WORLDTIME_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["WORLDTIME_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: rb/test/ipn_entity_test.rb
# Ipn entity test

require "minitest/autorun"
require "json"
require_relative "../WorldTime_sdk"
require_relative "runner"

class IpnEntityTest < Minitest::Test
  def test_create_instance
    testsdk = WorldTimeSDK.test(nil, nil)
    ent = testsdk.Ipn(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = ipn_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["load"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "ipn." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set WORLD_TIME_TEST_IPN_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    ipn_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.ipn")))
    ipn_ref01_data = nil
    if ipn_ref01_data_raw.length > 0
      ipn_ref01_data = Helpers.to_map(ipn_ref01_data_raw[0][1])
    end

    # LOAD
    ipn_ref01_ent = client.Ipn(nil)
    ipn_ref01_match_dt0 = {}
    ipn_ref01_data_dt0_loaded = ipn_ref01_ent.load(ipn_ref01_match_dt0, nil)
    assert !ipn_ref01_data_dt0_loaded.nil?

  end
end

def ipn_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "ipn", "IpnTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = WorldTimeSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["ipn01", "ipn02", "ipn03", "ip01", "ip02", "ip03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["WORLD_TIME_TEST_IPN_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "WORLD_TIME_TEST_IPN_ENTID" => idmap,
    "WORLD_TIME_TEST_LIVE" => "FALSE",
    "WORLD_TIME_TEST_EXPLAIN" => "FALSE",
  })

  idmap_resolved = Helpers.to_map(
    env["WORLD_TIME_TEST_IPN_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["WORLD_TIME_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
      },
      extra || {},
    ])
    client = WorldTimeSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["WORLD_TIME_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["WORLD_TIME_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
